src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java

```diff
@@ -0,0 +1,522 @@
+     */
+    public void setReallocation(ClusterService clusterService, AllocationService allocationService) {
+        this.clusterService = clusterService;
+        this.allocationService = allocationService;
+        this.clusterService.add(new ClusterStateListener() {
+            @Override
+            public void clusterChanged(ClusterChangedEvent event) {
+                DiscoveryNode localNode = event.state().nodes().localNode();
+                if (localNode != null) {
+                    if (localNode.masterNode() == true && event.localNodeMaster() == false) {
+                        cleanupOnNoMaster();
+                    }
+                } else {
+                    cleanupOnNoMaster();
+                }
+            }
+        });
+    }
+
+    /**
+     * When there is no master, we create a new generation of the cache, representing a new
+     * state that needs to be fetched. All ongoing fetch operations when done will make
+     * sure they are still working on the same cache generation, and if not, will not do
+     * anything since they are not relevant anymore.
+     */
+    // visible for testing
+    synchronized void cleanupOnNoMaster() {
+        cache.clear();
+        cacheGeneration.incrementAndGet();
+    }
+
+    /**
+     * Started shards just cause the cache to clean, so we can start fresh for another copy
+     * of the shard
+     */
+    public synchronized void applyStartedShards(StartedRerouteAllocation allocation) {
+        for (ShardRouting shardRouting : allocation.startedShards()) {
+            cache.remove(shardRouting.shardId());
+        }
+    }
+
+    /**
+     * Failed shards are tracked as if for the specific node, a fetch failure happened. At least for the
+     * first round of fetching data. The reason we need to keep it around is that now fetching is async,
+     * and we need to make sure we won't try and allocate again and again to the same node all the time.
+     */
+    public synchronized void applyFailedShards(FailedRerouteAllocation allocation) {
+        for (ShardRouting failedShard : allocation.failedShards()) {
+            Map<String, NodeEntry<T>> shardCache = cache.get(failedShard.shardId());
+            if (shardCache != null) {
+                // always mark the entry for that node as failed, to make sure we capture
+                // it next time we fetch the data
+                NodeEntry<T> nodeEntry = new NodeEntry<>(failedShard.currentNodeId());
+                nodeEntry.markAsFetching();
+                nodeEntry.doneFetching(new ElasticsearchException("marker for shard failed"));
+                shardCache.put(failedShard.currentNodeId(), nodeEntry);
+            }
+        }
+    }
+
+    /**
+     * Returns the number of async fetches that are currently ongoing.
+     */
+    public synchronized int getNumberOfAsyncFetches() {
+        int count = 0;
+        for (Map<String, NodeEntry<T>> shard : cache.values()) {
```

Comment:
it is calculated on the fly when calling this method. Do you mean keep it counted as things change? I prefer to keep it simple here since its not heavy calculation and its only really relevant during the heavy allocation phase of the cluster